### PR TITLE
[feat] ML-DSA87 driver

### PR DESCRIPTION
**Pull request by ArthurHeymans**
This updates the sw-emulator and adds a driver.

### CODE REVIEW

File: drivers/src/mldsa87.rs [hunk 1]
```diff
@@ -98,12 +155,46 @@ impl MlDsa87 {
...
+
+        // Program the command register for key generation
+        mldsa.ctrl().write(|w| w.ctrl(|w| w.keygen_sign()));
+
+        // Wait for hardware ready
+        MlDsa87::wait(mldsa, || mldsa.status().read().valid())?;
+
+        // Copy signature
+        let signature = MlDsa87Signature::read_from_reg(mldsa.signature());
+
```
Comment (mhatrevi) [inline drivers/src/mldsa87.rs:193]: Need to verify the signature. Hence the need for the pub_key.

File: drivers/src/mldsa87.rs [hunk 2]
```diff
@@ -98,12 +153,46 @@ impl MlDsa87 {
...
+        mldsa.ctrl().write(|w| w.zeroize(true));
+
+        // Copy seed from keyvault
+        KvAccess::copy_from_kv(*seed, mldsa.kv_rd_seed_status(), mldsa.kv_rd_seed_ctrl())
+            .map_err(|err| err.into_read_seed_err())?;
+
+        // Copy digest
+        KvAccess::copy_from_arr(msg, mldsa.msg())?;
+
+        // Sign RND, TODO do we want deterministic?
```
Comment (mhatrevi) [inline drivers/src/mldsa87.rs:176]: We need deterministic, but we'll keep the API flexible. Caller will always send sign_rnd as zeroes.

**Review by mhatrevi — CHANGES_REQUESTED** [review]

**Review by mhatrevi — APPROVED** [review]

File: drivers/src/mldsa87.rs [hunk 3]
```diff
@@ -116,13 +227,88 @@ impl MlDsa87 {
...
+        // Copy digest
+        msg.write_to_reg(mldsa.msg());
+
+        // Copy pubkey
+        pub_key.write_to_reg(mldsa.pubkey());
+
+        // Copy signature
+        signature.write_to_reg(mldsa.signature());
+
+        // Program the command register for key generation
```
Comment (swenson) [inline drivers/src/mldsa87.rs:254]:
```suggestion
        // Program the command register for verification
```

File: sw-emulator/lib/periph/src/ml_dsa87.rs [hunk 4]
```diff
@@ -341,11 +349,15 @@ impl MlDsa87 {
...
 
-        self.verify_res
-            .iter_mut()
-            .for_each(|e| *e = if result { 1 } else { 0 });
+        if success {
+            self.verify_res
+                .copy_from_slice(&self.signature[..ML_DSA87_VERIFICATION_SIZE / 4]);
+        } else {
+            self.verify_res
+                .copy_from_slice(&[0; ML_DSA87_VERIFICATION_SIZE / 4]);
```
Comment (swenson) [inline sw-emulator/lib/periph/src/ml_dsa87.rs:359]: Might be easier to do?
```suggestion
            self.verify_res.fill(0)
```

File: drivers/src/mldsa87.rs [hunk 5]
```diff
@@ -14,56 +14,85 @@ Abstract:
...
-pub type MlDsa87PubKey = Array4x648;
+pub type Mldsa87PubKey = Array4x648;
 
 /// MLDSA-87 Signature
-pub type MlDsa87Signature = Array4x1157;
+pub type Mldsa87Signature = Array4x1157;
 
 /// MLDSA-87 Message (64 Bytes)
-pub type MlDsa87MsgScalar = Array4x16;
+pub type Mldsa87MsgScalar = Array4x16;
```
Comment (swenson) [inline drivers/src/mldsa87.rs:40]: "Scalar" is not really correct for ML-DSA. In ECDSA, they were the scalar (integer) coordinates of a point, but in ML-DSA, they're just data.

I'd suggest just `Mldsa87Msg` and `Mldsa87SignRnd` or maybe `Mldsa87Data` and `Mldsa87SignRndData`

**Review by swenson — APPROVED** [review]
LGTM, a few minor nits.

File: drivers/src/mldsa87.rs (hunk 5 above)
Comment (mhatrevi) [reply drivers/src/mldsa87.rs:40]: Updated.

File: drivers/src/mldsa87.rs (hunk 3 above)
Comment (mhatrevi) [reply drivers/src/mldsa87.rs:254]: Updated.

File: sw-emulator/lib/periph/src/ml_dsa87.rs (hunk 4 above)
Comment (mhatrevi) [reply sw-emulator/lib/periph/src/ml_dsa87.rs:359]: Updated.

**Pull request merged**